### ISSUE TITLE
Investigate PVS-Studio checks

### ISSUE THREAD

**Issue by dominikh** [dominikh/go-tools#624]
https://www.viva64.com/en/w/
Last update (of this issue): Sat 12 Oct 2019 03:53:09 PM CEST

# Consider implementing

# Coarsely filtered

- A compound assignment expression 'X += X + N' is suspicious. Consider inspecting it for a possible error. (V765)
- A part of conditional expression is always true/false if it is evaluated. (V560, V3063)
- A pattern was detected: A || (A && ...). The expression is excessive or contains a logical error. (V686, V3017)
- A suspicious expression 'A[B < C]'. Probably meant 'A[B] < C'. (V661)
- A value of variable is not modified. Consider inspecting the expression. It is possible that '1' should be present instead of '0'. (V684, V3060)
- A wrong variable is probably used as an index in the for statement. (V787)
- An abnormality within similar comparisons. It is possible that a typo is present inside the expression. (V722)
- An excessive check can be simplified. The '||' operator is surrounded by opposite expressions 'x' and '!x'. (V728, V3031)
- An excessive expression. Examine the substrings "abc" and "abcd". (V734)
- An excessive type cast or check. The object is already of the same type. (V3051)
- An item with the same key has already been added. (V6033, V3058)
- An odd expression inside parenthesis. It is possible that a function name is missing. (V747)
- An odd precise comparison. Consider using a comparison with defined precision: Math.Abs(A - B) < Epsilon or Math.Abs(A - B) > Epsilon. (V550, V3024)
- An odd sequence of assignments of this kind: A = B; B = A; (V587, V3037)
- An unconditional 'break/continue/return/goto' within a loop. (V612, V3020, V6037)
- Be advised that compiler may delete this cycle or make it infinity. Use volatile variable(s) or synchronization primitives to avoid this. (V712)
- Be advised that the utilized constant value is represented by an octal form. (V536)
- Comparing objects of incompatible types. (V6058)
- Comparison with 'double.NaN' is meaningless. Use 'double.IsNaN()' method instead. (V3076, V6038)
- Compound assignment expression is used inside condition. (V1019)
- Consider inspecting an odd expression. Non-null function pointer is compared to null. (V516)
- Consider inspecting iterators which are being passed as arguments to function 'Foo'. (V539)
- Consider inspecting the 'for' operator. Initial and final values of the iterator are the same. (V625, V3028, V6043)
- Consider inspecting the 'for' operator. It's possible that the loop will be executed incorrectly or won't be executed at all. (V621, V3116)
- Consider inspecting the 'for' operator. No more than one iteration of the loop will be performed. (V1008)
- Consider inspecting the NN argument of the 'Foo' function. It is odd that the argument is of the 'T' type. (V632)
- Consider inspecting the application's logic. It's possible that 'else' keyword is missing. (V646, V3018)
- Consider inspecting the condition. An argument of the '|' bitwise operation always contains a non-zero value. (V617)
- Consider inspecting the condition. The 'Foo' pointer is always not equal to NULL. (V600)
- Consider inspecting the correctness of handling the N argument in the 'Foo' function. (V2013)
- Consider inspecting the expression for function call. It is possible that one of the closing ')' parentheses was positioned incorrectly. (V639)
- Consider inspecting the expression. Bit shifting of the 32-bit value with a subsequent expansion to the 64-bit type. (V629)
- Consider inspecting the expression. Probably the '!=' should be used here. (V3034)
- Consider inspecting the expression. Probably the '!=' should be used here. (V633)
- Consider inspecting the expression. Probably the '!='/'-='/'+=' should be used here. (V6015)
- Consider inspecting the expression. Probably the '+=' should be used here. (V3035)
- Consider inspecting the expression. Probably the '-=' should be used here. (V3036)
- Consider inspecting the loop expression. Different containers are utilized for setting up initial and final values of the iterator. (V662)
- Consider inspecting the loop expression. It is possible that different variables are used inside initializer and iterator. (V683, V3113)
- Consider inspecting this expression. '<' possibly should be replaced with '<<'. (V602)
- Consider inspecting this expression. The expression is excessive or contains a misprint. (V590, V3023, V6057)
- Consider reviewing the correctness of 'X' item's usage. (V537)
- Consider verifying the expression. An unsigned value is compared to the number - NN. (V605)
- Constant expression in switch statement. (V785, V3136, V6018)
- Dangerous magic number N used. (V112)
- Dereferencing of the null pointer might take place. (V522)
- Dividing sizeof a pointer by another value. There is a probability of logical error presence. (V514)
- Division or mod division by zero. (V609, V3064, V6020)
- Do not use real-type variables as loop counters. (V1034)
- Don't use terminating functions in library code. (V2014)
- Double negation is present in the expression: !!x. (V6035)
- Empirical analysis. It is possible that a typo is present inside the string literal. The 'foo' word is suspicious. (V691, V3091)
- Explicit conversion from 'float/double' type to signed integer type. (V2003)
- Explicit conversion from 'float/double' type to unsigned integer type. (V2004)
- Expression inside assert statement can change object's state. (V6055)
- Expression is always true/false. (V547, V3022, V6007)
- Function exited without performing epilogue actions. It is possible that there is an error. (V1020)
- Function receives an odd argument. (V575, V3057, V6009)
- Identical expressions to the left and to the right of compound assignment. (V3107, V6028)
- Incorrect format. Consider checking the N actual argument of the 'Foo' function. (V576, V3025, V6046)
- It is dangerous to create a local variable within a loop with a same name as a variable controlling this loop. (V711)
- It is likely that a wrong variable is being compared inside the 'for' operator. Consider reviewing 'X'. (V534, V3015)
- It is likely that a wrong variable is being incremented inside the 'for' operator. Consider reviewing 'X'. (V533, V3014)
- It is odd that a sizeof() operator is multiplied by sizeof(). (V531)
- It is odd that pointer is compared with the 'zero' value. Probably meant: *ptr != zero. (V528)
- It is odd that the 'zero' value is assigned to pointer. Probably meant: *ptr = zero. (V527)
- It is odd that the body of 'Foo_1' function is fully equivalent to the body of 'Foo_2' function. (V524, V3013, V6032)
- It is odd that the result of the statement is a part of the condition. Perhaps, this statement should have been compared with something else. (V793)
- It is pointless to compute the distance between the elements of different arrays. (V782)
- It is possible that ',' comma is missing at the end of the string. (V737)
- It is possible that 'else' block was forgotten or commented out, thus altering the program's operation logics. (V705, V3067)
- It is possible that the 'swap' function interchanges a variable with itself. (V671)
- It is possible that this 'else' branch must apply to the previous 'if' statement. (V563, V3033, V6047)
- It's dangerous to call the 'Foo' function in such a manner, as the line being passed could contain format specification. The example of the safe code: printf("%s", str); (V618)
- It's odd that this function always returns one and the same value of NN. (V657, V3009, V6014)
- It's odd to compare a bool type value with a value of N. (V562)
- It's possible that the line was commented out improperly, thus altering the program's operation logics. (V628, V3069)
- It's probably better to assign value to 'foo' variable than to declare it anew. (V561)
- Label is present inside a switch(). It is possible that these are misprints and 'default:' operator should be used instead. (V577)
- More than N bits are required to store the value, but the expression evaluates to the T type which can only hold K bits. (V673)
- NN identical blocks were found. (V761)
- Odd semicolon ';' after 'if/for/while' operator. (V529, V3007, V6063)
- Odd use of special character in regular expression. Possibly, it was intended to be escaped. (V6059)
- Parameter 'A' is always rewritten in method body before being used. (V763, V3061, V6023)
- Parameter is not used inside method's body. (V6022)
- Possible incorrect order of arguments passed to function. (V764, V3066, V6029)
- Possible infinite recursion. (V3110, V6062)
- Possible null dereference. (V3080)
- Possible overflow. Consider casting operands, not the result. (V1028)
- Possible use of a left shift operator instead of a comparison operator. (V770)
- Possibly index is out of bound. (V3106, V6025)
- Postfix increment/decrement is meaningless because this variable is overwritten. (V3133, V6044)
- Potential null dereference. (V6008)
- Potentially infinite loop. The variable in the loop exit condition does not change its value between iterations. (V776, V3120)
- Potentially unsafe double-checked locking. (V1036, V3054)
- Priority of the '&&' operation is higher than that of the '||' operation. (V648, V3130)
- Range intersections are possible within conditional expressions. (V695, V3092)
- Recurring check. This condition was already verified in previous line. (V571, V3030)
- Review captured variable in lambda expression. (V788)
- Saving the function result inside the 'byte' type variable is inappropriate. The significant bits could be lost breaking the program's logic. (V642)
- Shift by N bits is greater than the size of type. (V3134, V6034)
- Suspicious access to element by a constant index inside a loop. (V767, V3102, V6016)
- Suspicious subexpression in a sequence of similar comparisons. (V1013)
- The '&=' operation always sets a value of 'Foo' variable to zero. (V753)
- The 'X' counter is not used inside a nested loop. Consider inspecting usage of 'Y' counter. (V756, V3081, V6017)
- The 'first' argument of 'Foo' function is equal to the 'second' argument. (V549)
- The 'strcmp' function returns 0 if corresponding strings are equal. Consider examining the condition for mistakes. (V526)
- The 'then' statement is equivalent to the 'else' statement. (V523, V3004, V6004)
- The 'while' operator has empty body. Suspicious pattern detected. (V715)
- The 'x' variable is assigned to itself. (V3005, V6005)
- The 'x' variable is assigned values twice successively. Perhaps this is a mistake. (V519, V3008)
- The argument was passed to method several times. It is possible that another argument should be passed instead. (V3038)
- The code contains the collection of similar blocks. Check items X, Y, Z, ... in lines N1, N2, N3, ... (V525)
- The code under this 'case' label is unreachable. (V551)
- The condition of loop is always true/false. (V654)
- The conditional expressions of the 'if' statements situated alongside each other are identical. (V581, V3029)
- The constant NN is being utilized. The resulting value could be inaccurate. Consider using the M_NN constant from <math.h>. (V624)
- The expression is always false. Overflow check is incorrect. (V1012)
- The expression of 'foo(foo(x))' pattern is excessive or contains an error. (V754)
- The expression of the 'A - B > 0' kind will work as 'A != B'. (V555)
- The initial value of the index in the nested loop equals 'i'. Consider using 'i + 1' instead. (V791, V3135)
- The line contains control character 0x0B (vertical tabulation). (V538)
- The modification of a variable is unsequenced relative to another operation on the same variable. This may lead to undefined behavior. (V567)
- The object was used after it was verified against null. Check lines: N1, N2. (V3125)
- The object was used before it was verified against null. Check lines: N1, N2. (V3095)
- The operation is executed 2 or more times in succession. (V652, V3075)
- The operator evaluates both operands. Perhaps a short-circuit operator should be used instead. (V3093)
- The pointer in the expression equals nullptr. The resulting value is meaningless and should not be used. (V769)
- The pointer was used unsafely after it was verified against nullptr. (V1004)
- The pointer was utilized before it was verified against nullptr. Check lines: N1, N2. (V595, V713)
- The return value of function 'Foo' is required to be utilized. (V530, V3010, V6010)
- The same value is present on both sides of the operator. The expression is incorrect or it can be simplified. (V584)
- The switch statement does not cover all values of the enum. (V719, V3002, V6002)
- The use of 'if (A) {...} else if (A) {...}' pattern was detected. There is a probability of logical error presence. (V517, V3003, V6003)
- The used constant value is represented by an octal form. (V6061)
- The value is out of range of enum values. This causes unspecified or undefined behavior. (V1016)
- The value of the captured variable will be overwritten on the next iteration of the loop in each instance of anonymous function that captures it. (V3129)
- The value of the variable is checked after it was used. Perhaps there is a mistake in program logic. Check lines: N1, N2. (V781)
- The values of different enum types are compared. (V556)
- The variable 'X' is being used for this loop and for the outer loop. (V535, V3016, V6031)
- The variable is assigned but is not used by the end of the function. (V1001, V3137)
- The variable is assigned the same value on several loop iterations. (V1021)
- The variable is assigned to itself. (V570)
- The variable is incremented in the loop. Undefined behavior will occur in case of signed integer overflow. (V1026)
- The variable was utilized in the logical expression before it was verified against null in the same logical expression. (V3027)
- There are identical sub-expressions to the left and to the right of the 'foo' operator. (V501, V3001, V6001)
- There are two 'if' statements with identical conditional expressions. The first 'if' statement contains function return. This means that the second 'if' statement is senseless. (V649, V3021, V6039)
- This expression can be simplified. One of the operands in the operation equals NN. Probably it is a mistake. (V2007, V6048)
- This is a nonsensical comparison: pointer < 0. (V503)
- This value is already assigned to the 'b' variable. (V6026)
- Truncation of constant value. (V569)
- Two identical text blocks detected. The second block starts with NN string. (V760)
- Two opposite conditions were encountered. The second condition is always false. (V637, V3011)
- Two or more case-branches perform the same actions. (V1037)
- Two similar code fragments were found. Perhaps this is a typo and 'X' variable should be used instead of 'Y'. (V778, V3127)
- Unreachable code detected. It is possible that an error is present. (V779, V6019)
- Unsafe locking on an object. (V3090)
- Uppercase (lowercase) string is compared with a different lowercase (uppercase) string. (V3122)
- Usage of a suspicious mutex wrapper. It is probably unused, uninitialized, or already locked. (V1018)
- Verifying that a pointer value is not NULL is not required. The 'if (ptr != NULL)' check can be removed. (V809)
- Waiting on this expression is unreliable, as compiler may optimize some of the variables. Use volatile variable(s) or synchronization primitives to avoid this. (V3032)
- strcmp()-like functions can return not only the values -1, 0 and 1, but any values. (V698)

# Already implemented

# Not applicable to Go

- 'GetObjectData' implementation in unsealed type is not virtual, incorrect serialization of derived type is possible. (V3104)
- 'ThreadStatic' attribute is applied to a non-static 'A' field and will be ignored. (V3079)
- 'this == 0' comparison should be avoided - this comparison is always false on newer compilers. (V704)
- A 'break' statement is probably missing in a 'switch' statement. (V796)
- A 'wchar_t *' type string is incorrectly converted to 'BSTR' type string. (V745)
- A 32-bit variable is utilized as a reference to a pointer. A write outside the bounds of this variable may occur. (V207)
- A bool type variable is being incremented. Perhaps another variable should be incremented instead. (V552)
- A call of the 'Foo' function will lead to a buffer overflow or underflow. (V512)
- A class, containing pointers, constructor and destructor, is copied by the automatically generated operator= or copy constructor. (V1002)
- A code fragment from 'file' cannot be analyzed. (V001)
- A component of TimeSpan is used, which does not represent full time interval. Possibly 'Total*' value was intended instead. (V3118)
- A critical error had occurred. (V052)
- A dangerous cast of 'this' to 'void*' type in the 'Base' class, as it is followed by a subsequent cast to 'Class' type. (V725)
- A non-serializable class should not be serialized. (V6065)
- A number of elements in the allocated array is equal to size of a pointer in bytes. (V697)
- A pointer without owner is added to the container by the 'emplace_back' method. A memory leak will occur in case of an exception. (V1023)
- A private Ctor(SerializationInfo, StreamingContext) constructor in unsealed type will not be accessible when deserializing derived types. (V3103)
- A suspicious function declaration. It is possible that the T type object was meant to be created. (V644)
- A suspicious string consisting of two parts is used for the initialization. It is possible that a comma is missing. (V653)
- A terminal null is present inside a string. '\0xNN' character sequence was encountered. Probably meant: '\xNN'. (V3132)
- A terminal null is present inside a string. The '\0xNN' characters were encountered. Probably meant: '\xNN'. (V638)
- A variable of the memsize type is read from a stream. Consider verifying the compatibility of 32 and 64 bit versions of the application in the context of a stored data. (V128)
- Allocation of memory by the pattern "(X*)malloc(sizeof(Y))" where the sizes of X and Y types are not equal. (V123)
- An array is being utilized as a pointer to single object. (V619)
- An array/object was declared but was not utilized. (V808)
- An attempt to free memory containing the 'int A[10]' array by using the 'free(A)' function. (V726)
- An attempt to release the memory in which the 'Foo' local variable is stored. (V585)
- An empty exception handler. Silent suppression of exceptions can hide the presence of bugs in source code during testing. (V565)
- An enumeration was declared with 'Flags' attribute, but no initializers were set to override default values. (V3121)
- An exception was thrown by pointer. Consider throwing it by value instead. (V1022)
- An excessive expression. Examine the substrings "abc" and "abcd". (V3053)
- An inappropriate attempt to append a null character to a string. To determine the length of a string by 'strlen' function correctly, a string ending with a null terminator should be used in the first place. (V692)
- An object is used as an argument to its own method. Consider checking the first actual argument of the 'Foo' function. (V678)
- An object is used as an argument to its own method. Consider checking the first actual argument of the 'Foo' method (V3062)
- An odd explicit conversion from 'float *' type to 'double *' type. (V615)
- An odd implicit type casting. (V601)
- An odd operation of the 'sizeof(X)/sizeof(T)' kind is performed, where 'X' is of the 'class' type. (V651)
- An overflow of the 32-bit variable is possible inside a long cycle which utilizes a memsize-type loop counter. (V127)
- An uninitialized class member is used to initialize another member. Remember that members are initialized in the order of their declarations inside a class. (V670)
- Analysis of 'Makefile/Utility' type projects is not supported in this tool. Use direct analyzer integration or compiler monitoring instead. (V010)
- Anonymous function is used to unsubscribe from event. No handlers will be unsubscribed, as a separate delegate instance is created for each anonymous function declaration. (V3084)
- Array overrun is possible. (V557)
- BSTR string becomes invalid. Notice that BSTR strings store their length before start of the text. (V750)
- Be advised that the size of the type 'long' varies between LLP64/LP64 data models. (V126)
- Because NULL is defined as 0, the exception is of the 'int' type. Keyword 'nullptr' could be used for 'pointer' type exception. (V740)
- C-style explicit type casting is utilized. Consider using: static_cast/const_cast/reinterpret_cast. (V2005)
- Call of function 'foo' with variable number of arguments. N argument has memsize type. (V111)
- Calling a virtual (overridden) event may lead to unpredictable behavior. Consider implementing event accessors explicitly or use 'sealed' keyword. (V3119)
- Calling an overridden method in parent-class constructor may lead to use of uninitialized data. (V6052)
- Calling overrideable class member from constructor is dangerous. (V3068)
- Calling the 'delete' operator for a void pointer will cause undefined behavior. (V772)
- Cannot determine active configuration for project. Please check projects and solution configurations. (V005)
- Check the array initialization. Only the first element is initialized explicitly. (V1009)
- Checking value for null will always return false when generic type is instantiated with a value type. (V3111)
- Class initialization cycle is present. (V6050)
- Classes should always be derived from std::exception (and alike) as 'public'. (V702)
- Classes should not be compared by their name. (V6054)
- Classes that define 'equals' method must also define 'hashCode' method. (V6049)
- Collection is modified while iteration is in progress. ConcurrentModificationException may occur. (V6053)
- Comparison of arrays, strings, collections by reference. Possibly an equality comparison was intended. (V6013)
- Consider adding '[Flags]' attribute to the enum. (V3059)
- Consider inspecting an odd type cast: 'Type1' to ' Type2'. (V542)
- Consider inspecting conditional expression of the loop. It is possible that 'i < X.size()' should be used instead of 'X.size()'. (V693)
- Consider inspecting signed and unsigned function arguments. See NN argument of function 'Foo' in derived class and base class. (V2011)
- Consider inspecting the '?:' operator. A temporary object is being created and subsequently destroyed. (V623)
- Consider inspecting the 'T foo = foo = x;' expression. It is odd that variable is initialized through itself. (V700)
- Consider inspecting the 'foo = bar = baz ? .... : ....' expression. It is possible that 'foo = bar == baz ? .... : ....' should be used here instead. (V699)
- Consider inspecting the 'switch' statement. It's possible that the first 'case' operator is missing. (V622)
- Consider inspecting the expression. The length should probably be multiplied by the sizeof(wchar_t). (V635)
- Consider inspecting the statement of '*pointer++' pattern. Probably meant: '(*pointer)++'. (V532)
- Consider inspecting virtual function arguments. See NN argument of function 'Foo' in derived class and base class. (V762)
- Consider passing the 'Foo' argument as a constant pointer/reference. (V2009)
- Consider reviewing type casting. TYPE X[][] is not equivalent to TYPE **X. (V548)
- Constructor parameter is not used. (V3117)
- Converting integers or pointers to BOOL can lead to a loss of high-order bits. Non-zero value can become 'FALSE'. (V724)
- Copying from unsafe data source. Buffer overflow is possible. (V755)
- Creating an object with placement new requires a buffer of large size. (V752)
- Custom declaration of a standard type. The declaration from system header files should be used instead. (V677)
- Dangerous construction is used: 'm[x] = m.size()', where 'm' is of 'T' class. This may lead to undefined behavior. (V708)
- Dangerous widening type conversion from an array of derived-class objects to a base-class pointer. (V777)
- Declarations of functions with 'Foo' name differ in the 'const' keyword only, but the bodies of these functions have different composition. This is suspicious and can possibly be an error. (V659)
- Decreased performance. Consider creating a pointer/reference to avoid using the same expression repeatedly. (V807)
- Decreased performance. Consider replacing the expression 'AA' with 'BB'. (V815)
- Decreased performance. Excessive type casting: string -> char * -> string. (V811)
- Decreased performance. Ineffective use of the 'count' function. It can possibly be replaced by the call to the 'find' function. (V812)
- Decreased performance. It is better to redefine the N function argument as a reference. Consider replacing 'const T' with 'const .. &T' / 'const .. *T'. (V801)
- Decreased performance. It is inefficient to identify an empty string by using 'strlen(str) > 0' construct. A more efficient way is to check: str[0] != '\0' (V805)
- Decreased performance. It is more effective to use the prefix form of ++it. Replace iterator++ with ++iterator. (V803)
- Decreased performance. Memory is allocated and released multiple times inside the loop body. (V819)
- Decreased performance. The 'Foo' function is called twice in the specified expression to calculate length of the same string. (V804)
- Decreased performance. The 'strlen' function was called multiple times inside the body of a loop. (V814)
- Decreased performance. The argument should probably be rendered as a constant pointer/reference. (V813)
- Decreased performance. The expression of strlen(MyStr.c_str()) kind can be rewritten as MyStr.length(). (V806)
- Deprecated CLR switch was detected. Incorrect diagnostics are possible. (V007)
- Dereferencing of invalid iterator 'X' might take place. (V783)
- Destructor of the object will be invoked a second time after leaving the object's scope. (V749)
- Diagnostics from the 64-bit rule set are not entirely accurate without the appropriate 64-bit compiler. Consider utilizing 64-bit compiler if possible. (V004)
- EOF should not be compared with a value of the 'char' type. Consider using the 'int' type. (V739)
- Explicit conversion from 'void *' to 'int *'. (V206)
- Explicit conversion from 32-bit integer type to memsize type. (V201)
- Explicit conversion from 32-bit integer type to pointer type. (V204)
- Explicit conversion from memsize type to 32-bit integer type. (V202)
- Explicit conversion of pointer type to 32-bit integer type. (V205)
- Explicit type conversion from memsize to double type or vice versa. (V203)
- File cannot be processed. Analysis aborted by timeout. (V006)
- Function 'Foo' writes/reads 'N' bytes. The alignment rules and type sizes have been changed. Consider reviewing this value. (V124)
- Function body contains the 'X' label that is not used by any 'goto' statements. (V729)
- Function execution could be deferred. Consider specifying execution policy explicitly. (V1011)
- Function is not declared. The passing of data to or from this function may be affected. (V1031)
- Function receives an address of a 'char' type variable instead of pointer to a buffer. (V742)
- Function returns a pointer to the internal string buffer of a local object, which will be destroyed. (V723)
- Function returns the pointer/reference to temporary local object. (V558)
- Handling of two different exception types is identical. (V2010)
- IDisposable object is not disposed before method returns. (V3114)
- Implementation of 'compareTo' overloads the method from a base class. Possibly, an override was intended. (V6056)
- Implicit assignment type conversion to memsize type. (V101)
- Implicit conversion of the type of 'new' operator's argument to size_t type. (V121)
- Implicit type conversion N argument of function 'foo' to 32-bit type. (V107)
- Implicit type conversion N argument of function 'foo' to memsize type. (V106)
- Implicit type conversion from enum type to integer type. (V2006)
- Implicit type conversion from memsize to double type or vice versa. (V113)
- Implicit type conversion from memsize type to 32-bit type. (V103)
- Implicit type conversion of return value from memsize type to 32-bit type. (V110)
- Implicit type conversion of return value to memsize type. (V109)
- Implicit type conversion to memsize type in an arithmetic expression. (V104)
- Incorrect index type: 'foo[not a memsize-type]'. Use memsize type instead. (V108)
- Incorrect use of smart pointer. (V554)
- Infinite loop is possible. The 'cin.eof()' condition is insufficient to break from the loop. Consider adding the 'cin.fail()' function call to the conditional expression. (V663)
- Initializer of a field marked by [ThreadStatic] attribute will be called once on the first accessing thread. The field will have default value on different threads. (V3089)
- It is advised to utilize the 'SuspendThread' function only when developing a debugger (see documentation for details). (V720)
- It is dangerous to print a string into itself. (V541)
- It is highly probable that the semicolon ';' is missing after 'return' keyword. (V504)
- It is incorrect to compare the variable of BOOL type with TRUE. (V676)
- It is more efficient to catch exception by reference rather than by value. (V816)
- It is more efficient to search for 'X' character rather than a string. (V817)
- It is more efficient to use an initialization list rather than an assignment operator. (V818)
- It is not advised to declare type 'T' as 32-bit type. (V125)
- It is not recommended to return null or throw exceptions from 'ToString()' method. (V3108)
- It is not recommended to throw exceptions from 'Equals(object obj)' method. (V3115)
- It is odd that the 'foo' field in derived class overwrites field in base class. (V703)
- It is odd that the BSTR data type is compared using a relational operator. (V775)
- It is odd that the assignment operator takes an object by a non-constant reference and returns this object. (V790)
- It is odd that the object which was created using 'new' operator is immediately cast to another type. (V572)
- It is odd that the value 'X' of HRESULT type is compared with 'Y'. (V544)
- It is odd that value 'X' is assigned to the variable 'Y' of HRESULT type. (V543)
- It is possible that an incorrect variable is compared with null after type conversion using 'as' keyword. (V3019)
- It is possible that an incorrect variable is compared with null after type conversion using 'dynamic_cast'. (V757)
- It is possible that macro expansion resulted in incorrect evaluation order. (V733)
- It is suspicious to cast object of base class V to derived class U. (V717)
- It's odd that a char or string literal is added to a pointer. (V1038)
- It's odd that the argument of sizeof() operator is the expression. (V568)
- Iterators for the container, used in the range-based for loop, become invalid upon a function call. (V789)
- MISRA. A function should have a single point of exit at the end. (V2506)
- MISRA. A switch-expression should not have Boolean type. Consider using of 'if-else' construct. (V2527)
- MISRA. A switch-label should only appear at the top level of the compound statement forming the body of a 'switch' statement. (V2524)
- MISRA. All integer constants of unsigned type should have 'u' or 'U' suffix. (V2523)
- MISRA. Any label should be declared in the same block as 'goto' statement or in any block enclosing it. (V2529)
- MISRA. Any loop should be terminated with no more than one 'break' or 'goto' statement. (V2530)
- MISRA. Arrays should not be partially initialized. (V2540)
- MISRA. C-style and functional notation casts should not be performed. (V2533)
- MISRA. Class destructor should not exit with an exception. (V2539)
- MISRA. Declaration should contain no more than two levels of pointer nesting. (V2515)
- MISRA. Every 'switch' statement should contain non-empty switch-clauses. (V2525)
- MISRA. Every 'switch' statement should have a 'default' label, which, in addition to the terminating 'break' statement, should contain either a statement or a comment. (V2519)
- MISRA. Every switch-clause should be terminated by an unconditional 'break' or 'throw' statement. (V2520)
- MISRA. Expression of essential type 'foo' should not be explicitly cast to essential type 'bar'. (V2531)
- MISRA. Function should not be declared implicitly. (V2541)
- MISRA. Function should not contain labels not used by any 'goto' statements. (V2536)
- MISRA. Function with a non-void return type should return a value from all exit paths. (V2542)
- MISRA. Functions should not have unused parameters. (V2537)
- MISRA. Implicitly specified enumeration constants should be unique – consider specifying non-unique constants explicitly. (V2503)
- MISRA. Literal suffixes should not contain lowercase characters. (V2517)
- MISRA. Memory allocation and deallocation functions should not be used. (V2511)
- MISRA. Octal constants should not be used. (V2501)
- MISRA. Only the first member of enumerator list should be explicitly initialized, unless all members are explicitly initialized. (V2521)
- MISRA. Size of an array is not specified. (V2504)
- MISRA. String literal should not be assigned to object unless it has type of pointer to const-qualified char. (V2532)
- MISRA. The 'default' label should be either the first or the last label of a 'switch' statement. (V2518)
- MISRA. The 'goto' statement should not be used. (V2502)
- MISRA. The 'goto' statement shouldn't jump to a label declared earlier. (V2505)
- MISRA. The 'if' ... 'else if' construct should be terminated with an 'else' statement. (V2516)
- MISRA. The 'switch' statement should have 'default' as the last label. (V2522)
- MISRA. The body of a loop\conditional statement should be enclosed in braces. (V2507)
- MISRA. The comma operator should not be used. (V2528)
- MISRA. The function with the 'abort/exit/getenv/system' name should not be used. (V2509)
- MISRA. The function with the 'atof/atoi/atoll/atoll' name should not be used. (V2508)
- MISRA. The function with the 'qsort/bsearch' name should not be used. (V2510)
- MISRA. The functions from time.h/ctime should not be used. (V2526)
- MISRA. The loop counter should not have floating-point type. (V2534)
- MISRA. The macro with the 'setjmp' name and the function with the 'longjmp' name should not be used. (V2512)
- MISRA. The value of uninitialized variable should not be used. (V2538)
- MISRA. Unbounded functions performing string operations should not be used. (V2513)
- MISRA. Unions should not be used. (V2514)
- MISRA. Unreachable code should not be present in the project. (V2535)
- Member 'x' should point to string terminated by two 0 characters. (V540)
- Member of a class is initialized with itself: 'Foo(Foo)'. (V546)
- Member operator[] of 'foo' class has a 32-bit type argument. Use memsize-type here. (V302)
- Member operator[] of object 'foo' is declared with 32-bit type argument, but is called with memsize type argument. (V120)
- Memory for 'getline' function should be allocated only by 'malloc' or 'realloc' functions. Consider inspecting the first parameter of 'getline' function. (V748)
- Memsize type is used for catch. (V116)
- Memsize type is used for throw. (V115)
- Memsize type is used in the struct/class. (V122)
- Memsize type is used in the union. (V117)
- More than one sizeof() operator is used in one expression. (V119)
- N operand of '?:' operation: implicit type conversion to memsize type. (V105)
- Non-void function should return a value. (V591)
- Not all IDisposable members are properly disposed. Call 'Dispose' when disposing 'A' class. (V3073)
- Not all members of a class are initialized inside the constructor. (V730)
- Not all the members of type are serialized inside 'GetObjectData' method. (V3099)
- NullReferenceException is possible. Unhandled exceptions in destructor lead to termination of runtime. (V3100)
- Numeric Truncation Error. Return value of function is written to N-bit variable. (V1029)
- Object slicing. An exception should be caught by reference rather than by value. (V746)
- Only values that are returned from fgetpos() can be used as arguments to fsetpos(). (V1035)
- Original sorting order will be lost after repetitive call to 'OrderBy' method. Use 'ThenBy' method to preserve the original sorting. (V3078)
- Ownerless expression 'Foo'. (V607)
- Ownerless token 'Foo'. (V606)
- Passing objects of incompatible types to the method of collection. (V6066)
- Perhaps the '?:' operator works in a different way than it was expected. The '?:' operator has a lower priority than the 'foo' operator. (V502)
- Perhaps the '??' operator works differently from what was expected. Its priority is lower than that of other operators in its left part. (V3123)
- Pointer is cast to a more strictly aligned pointer type. (V1032)
- Pointer to an object of the class is cast to unrelated class. (V1027)
- Pointer to local array 'X' is stored outside the scope of this array. Such a pointer will become invalid. (V507)
- Pointer to local variable 'X' is stored outside the scope of this variable. Such a pointer will become invalid. (V506)
- Possibility of decreased performance. It is advised to pass arguments to std::unary_function/std::binary_function template as references. (V2012)
- Possible NullReferenceException. The '?.' and '.' operators are used for accessing members of the same object. (V3042)
- Possible exception when deserializing type. The Ctor(SerializationInfo, StreamingContext) constructor is missing. (V3094)
- Possible exception when serializing type. [Serializable] attribute is missing. (V3096)
- Possible exception: type marked by [Serializable] contains non-serializable members not marked by [NonSerialized]. (V3097)
- Possibly an incorrect HTML. The "</XX" closing tag was encountered, while the "</YY" tag was expected. (V735)
- Possibly an incorrect HTML. The </XX> closing tag was encountered, while the </YY> tag was expected. (V3050)
- Possibly, the usage of '#pragma warning(default: X)' is incorrect in this context. The '#pragma warning(push/pop)' should be used instead. (V665)
- Potential resurrection of 'this' object instance from destructor. Without re-registering for finalization, destructor will not be called a second time on resurrected object. (V3101)
- Presence of #line directives may cause some diagnostic messages to have incorrect file name and line number. (V011)
- Property setter / event accessor does not utilize its 'value' parameter. (V3077)
- Rather than creating 'std::unique_lock' to lock on the mutex, a new variable with default value is created. (V1025)
- Recurring sequence of explicit type casts. (V608)
- Reference invalidated because of the destruction of the temporary object returned by the function. (V758)
- Return value of 'wcslen' function is not multiplied by 'sizeof(wchar_t)' (V727)
- Several shared_ptr objects are initialized by the same pointer. A double memory deallocation will occur. (V1006)
- Size of an array calculated by the sizeof() operator was added to a pointer. It is possible that the number of elements should be calculated by sizeof(A)/sizeof(A[0]). (V687)
- Some diagnostic messages may contain incorrect line number. (V002)
- Some of the references in project are missing or incorrect. The analysis results could be incomplete. Consider making the project fully compilable and building it before analysis. (V051)
- Strange pointer arithmetic with 'malloc/new'. (V613)
- Structures with members of real type are compared byte-wise. (V1014)
- Such conditional expression of 'if' statement is incorrect for the HRESULT type value 'Foo'. The SUCCEEDED or FAILED macro should be used instead. (V545)
- Such expressions using the ',' operator are dangerous. Make sure the expression is correct. (V521)
- Suspicious assignment inside the condition expression of 'if/while/for' operator. (V3055)
- Suspicious assignment inside the conditional expression of 'if/while/do...while' statement. (V6041)
- Suspicious assignment inside the conditional expression of 'if/while/for' statement. (V559)
- Suspicious comparison found: 'a == b == c'. Remember that 'a == b == c' is not equal to 'a == b && b == c'. (V709)
- Suspicious declaration found. There is no point to declare constant reference to a number. (V710)
- Suspicious division: sizeof(X) / Value. Size of every element in X array does not equal to divisor. (V706)
- Suspicious invocation of Thread.run(). (V6064)
- Suspicious sequence of types castings: memsize -> 32-bit integer -> memsize. (V220)
- Suspicious sequence of types castings: pointer -> memsize -> 32-bit integer. (V221)
- Suspicious simultaneous use of bitwise and logical operators. (V1015)
- Suspicious type conversion: HRESULT -> BOOL (BOOL -> HRESULT). (V716)
- Temporary anonymous object is used. (V738)
- Temporary object is immediately destroyed after being created. Consider naming the object. (V744)
- The '&' or '|' operator is applied to bool type value. You've probably forgotten to include parentheses or intended to use the '&&' or '||' operator. (V564)
- The '?:' operator uses constants from different enums. (V771)
- The '?:' operator, regardless of its conditional expression, always returns one and the same value. (V3012)
- The '?:' operator, regardless of its conditional expression, always returns one and the same value. (V583)
- The '?:' operator, regardless of its conditional expression, always returns one and the same value. (V6012)
- The 'A' class containing IDisposable members does not itself implement IDisposable. (V3072)
- The 'A' class contains 'Dispose' method. Consider making it implement 'IDisposable' interface. (V3074)
- The 'Foo' function is called twice for deallocation of the same resource. (V586)
- The 'Foo' function is not expected to receive class-type variable as 'N' actual argument. (V510)
- The 'Foo' function receives the pointer and its size as arguments. It is possibly a mistake. Inspect the N argument. (V579)
- The 'Foo' function should not be called from 'DllMain' function. (V718)
- The 'Thread' object is created but is not started. It is possible that a call to 'Start' method is missing. (V3082)
- The 'X' variable was not initialized. This variable is passed by a reference to the 'Foo' function in which its value will be utilized. (V679)
- The 'a' variable was used after it was assigned through null-conditional operator. NullReferenceException is possible. (V3105)
- The 'alloca' function is used inside the loop. This can quickly overflow stack. (V505)
- The 'continue' operator will terminate 'do { ... } while (FALSE)' loop because the condition is always false. (V696)
- The 'continue' operator will terminate 'do { ... } while (false)' loop because the condition is always false. (V3098)
- The 'continue' operator will terminate 'do { ... } while (false)' loop because the condition is always false. (V6024)
- The 'delete A, B' expression only destroys the 'A' object. Then the ',' operator returns a resulting value from the right side of the expression. (V680)
- The 'delete' operator is applied to non-pointer. (V515)
- The 'foo' local variable possesses the same name as one of the class members, which can result in a confusion. (V688)
- The 'malloc' function allocates strange amount of memory calculated by 'strlen(expr)'. Perhaps the correct variant is strlen(expr) + 1. (V518)
- The 'malloc' function is used to allocate memory for an array of objects which are classes containing constructors/destructors. (V630)
- The 'memset/memcpy' function is used to nullify/copy the fields of 'Foo' class. Virtual table pointer will be damaged by this. (V598)
- The 'throw' operator does not possess any arguments and is not situated within the 'catch' block. (V667)
- The 'throw' operator inside the destructor should be placed within the try..catch block. Raising exception inside the destructor is illegal. (V509)
- The VARIANT_BOOL type is utilized incorrectly. The true value (VARIANT_TRUE) is defined as -1. (V721)
- The argument is a non-constant reference. The analyzer is unable to determine the position at which this argument is being modified. It is possible that the function contains an error. (V669)
- The behavior is undefined for arithmetic or comparisons with pointers that do not point to members of the same array. (V736)
- The class implements a copy constructor/operator=, but lacks the operator=/copy constructor. (V690)
- The code's operational logic does not correspond with its formatting. (V3043)
- The code's operational logic does not correspond with its formatting. (V6040)
- The code's operational logic does not correspond with its formatting. (V640)
- The comma operator ',' in array index expression. (V520)
- The compiler could delete the 'memset' function call, which is used to flush 'Foo' buffer. The RtlSecureZeroMemory() function should be used to erase the private data. (V597)
- The condition (ptr - const_value) is only false if the value of a pointer equals a magic constant. (V694)
- The destructor of the 'Foo' class is not declared as a virtual. It is possible that a smart pointer will not destroy an object correctly. (V689)
- The expression contains a suspicious mix of integer and real types (V3040)
- The expression contains a suspicious mix of integer and real types (V674, V6011)
- The expression of the 'A =+ B' kind is utilized. Consider reviewing it, as it is possible that 'A += B' was meant. (V588)
- The expression of the 'A =- B' kind is utilized. Consider reviewing it, as it is possible that 'A -= B' was meant. (V589)
- The expression was implicitly cast from integer type to real type. Consider utilizing an explicit type cast to avoid overflow or loss of a fractional part. (V636)
- The expression was implicitly cast from integer type to real type. Consider utilizing an explicit type cast to avoid the loss of a fractional part. (V3041)
- The field (property) is used before it is initialized in constructor. (V3128)
- The following pattern is used: throw (a, b);. It is possible that type name was omitted: throw MyException(a, b);. (V741)
- The function call could lead to the buffer overflow. The bounds should not contain the size of the buffer, but a number of characters it can hold. (V645)
- The function is deprecated in the Win64 system. It is safer to use the 'foo' function. (V303)
- The function is used as if it returned a bool type. The return value of the function should probably be compared with std::string::npos. (V797)
- The function was exited without releasing the pointer/handle. A memory/resource leak is possible. (V773)
- The integer constant is converted to pointer. Possibly an error or a bad coding style. (V566)
- The macro expression is dangerous, or it is suspicious. (V1003)
- The memory areas must not overlap. Use 'memmove' function. (V743)
- The name of 'X' field/property in a nested type is ambiguous. The outer type contains static field/property with identical name. (V3085)
- The object is returned from inside 'using' block. 'Dispose' will be invoked before exiting method. (V3071)
- The object of non-passive (non-PDS) type cannot be used with the function. (V780)
- The object was created but it is not being used. If you wish to call constructor, 'this->Foo::Foo(....)' should be used. (V603)
- The object was created but it is not being used. The 'throw' keyword could be missing. (V3006)
- The object was created but it is not being used. The 'throw' keyword could be missing. (V596)
- The object was created but it is not being used. The 'throw' keyword could be missing. (V6006)
- The original exception object was swallowed. Stack of original exception could be lost. (V3052)
- The pointer is being dereferenced on the initialization list before it is verified against null inside the body of the constructor function. (V664)
- The pointer is used simultaneously as an array and as a pointer to single object. (V574)
- The pointer steps out of array's bounds. (V594)
- The pointer was used after the memory was released. (V774)
- The priority of the '+' operation is higher than that of the '<<' operation. It's possible that parentheses should be used in the expression. (V634)
- The program contains an unused label and a function call: 'CC:AA()'. It's possible that the following was intended: 'CC::AA()'. (V660)
- The reference was used before it was verified against null. (V6060)
- The resource was acquired using 'X' function but was released using incompatible 'Y' function. (V1005)
- The sizeof() operator returns size of the pointer, and not of the array, in given expression. (V511)
- The stream is checked for EOF before reading from it, but is not checked after reading. Potential use of invalid data. (V1024)
- The use of 'new type(n)' pattern was detected. Probably meant: 'new type[n]'. (V508)
- The value from the uninitialized optional is used. (V6036)
- The value from the uninitialized optional is used. Probably it is a mistake. (V1007)
- The value of 'A' type is assigned to the pointer of 'B' type. (V647)
- The variable is not used after copying. Copying can be replaced with move/swap for optimization. (V820)
- The variable is used after it was moved. (V1030)
- The variable of char type is compared with pointer to string. (V731)
- The virtual destructor is not present, although the 'Foo' class contains virtual functions. (V599)
- There is no sense in testing the pointer against null, as the memory was allocated using the 'new' operator. The exception will be generated in the case of memory allocation error. (V668)
- There is probably no need in creating a new variable here. One of the function's arguments possesses the same name and this argument is a reference. (V672)
- To use free version of PVS-Studio, source code files are required to start with a special comment. (V009)
- Type casting operation is utilized 2 times in succession. Next, the '+' operation is executed. Probably meant: (T1)((T2)a + b). (V650)
- Type implementing IEquatable<T> interface does not override 'GetHashCode' method. (V3126)
- Type of variable enumerated in 'foreach' is not guaranteed to be castable to the type of collection's elements. (V3087)
- Unable to start the analysis on this file. (V008)
- Unary minus operator does not modify a bool type value. (V732)
- Unchecked tainted data is used in expression. (V1010)
- Unexpected function overloading behavior. See N argument of function 'foo' in derived class 'derived' and base class 'base'. (V301)
- Uninitialized variable 'Foo' used. (V614)
- Uninitialized variable 'Foo' was used. The variable was used to initialize itself. (V573)
- Uninitialized variables are used when initializing the 'A' variable. (V3070)
- Unrecognized error found... (V003)
- Unsafe invocation of event, NullReferenceException is possible. Consider assigning event to a local variable before invoking it. (V3083)
- Unusual pointer arithmetic. The value of the 'char' type is being added to the string pointer. (V643)
- Usage of non memsize type for pointer arithmetic. (V102)
- Use _beginthreadex/_endthreadex functions instead of CreateThread/ExitThread functions. (V513)
- Use of jump statements in 'finally' block can lead to the loss of unhandled exceptions. (V6051)
- Variable is declared as auto in C. Its default type is int. (V1033)
- Variable is not passed into foreach loop by a reference, but its value is changed inside of the loop. (V714)
- Variable of the 'string_view' type references a temporary object which will be removed after evaluation of an expression. (V1017)
- Violated order of exception handlers. Exception caught by handler for base class. (V759)
- WPF: A class containing registered property does not correspond with a type that is passed as the ownerType.type. (V3047)
- WPF: readonly field of 'DependencyProperty' type is not initialized. (V3049)
- WPF: several Dependency Properties are registered with a same name within the owner type. (V3048)
- WPF: the names of the property registered for DependencyProperty, and of the property used to access it, do not correspond with each other. (V3045)
- WPF: the type registered for DependencyProperty does not correspond with the type of the property used to access it. (V3046)
- WPF: writing and reading are performed on a different Dependency Properties. (V3044)
- Writing into the read-only memory. (V675)
- malloc() function accepts a dangerous expression in the capacity of an argument. (V118)
- realloc() possible leak: when realloc() fails in allocating memory, original pointer is lost. Consider assigning realloc() to a temporary pointer. (V701)
- Note that the size of the 'time_t' type is not 64 bits. After the year 2038, the program will work incorrectly. (V795)
- Consider inspecting the expression. The argument of sizeof() is the macro which expands to a number. (V627)
- Consider inspecting the return statement. The expression contains a comma. (V685)
- The variable is not used after memory has been allocated for it. Consider checking the use of this variable. (V799)
- The variable is of enum type. It is odd that it is used as a variable of a Boolean-type. (V768)
- The expression is checked for compatibility with type 'A' but is cast to type 'B'. (V3131, V6042)
- The assignment operator should be protected from the case of this == &src. (V794)
- Undefined behavior. Check the shift operator. (V610)
- Dangerous explicit type pointer conversion. (V114)
- It is odd that the number of iterations in the loop equals to the size of the pointer. (V604)
- It's unusual that the expression of sizeof(T)*N kind is being summed with the pointer to T type. (V620)
- Consider reviewing the expression of the 'A = B == C' kind. The expression is calculated as following: 'A = (B == C)'. (V593)
- Appending an element and checking for key uniqueness is performed on two different variables. (V3124)
- Consider checking for misprints. It's possible that ',' should be replaced by ';'. (V626)
- Consider inspecting NN argument of the function 'Foo'. It is possible that the value does not correspond with the length of a string which was passed with the YY argument. (V666)

# Not desirable

- Consider inspecting the 'Foo' function call. Defining an absolute path to the file or directory is considered a poor style. (V3039)
- Cyclomatic complexity: NN. Consider refactoring the 'Foo' function. (V2008)
- Decreased performance. The 'A' function was called several times with identical arguments. The result should possibly be saved to a temporary variable, which then could be used while calling the 'B' function. (V810)
- Giving short names to global variables is considered to be bad practice. (V707)
- On 32-bit/64-bit platform, structure size can be reduced from N to K bytes by rearranging the fields according to their sizes in decreasing order. (V802)
- Parameter is not used inside function's body. (V751)
- Parameter is not utilized inside method's body. (V3065)
- The length of function's body or class's declaration is more than 2000 lines long. You should consider refactoring the code. (V553)
- The variable can be constructed in a lower level scope. (V821)
- Consider inspecting the 'Foo' function call. Defining an absolute path to the file or directory is considered a poor style. (V631)
- Suspicious literal is present: '/r'. It is possible that a backslash should be used here instead: '\r'. (V682)

# Not yet classified

- A value is being subtracted from the unsigned variable. This can result in an overflow. In such a case, the comparison operation can potentially behave unexpectedly. (V658)
- An abnormality within similar comparisons. It is possible that a typo is present inside the expression. (V3112)
- An item with the same key has already been added. (V766)
- An odd bitwise operation detected. Consider verifying it. (V578)
- An odd explicit type casting. Consider verifying it. (V580)
- Assigning the value C to the X variable looks suspicious. The value range of the variable: [A, B]. (V786)
- Consider reviewing the correctness of 'X' item's usage. (V3056)
- Consider reviewing the source code which operates the container. (V582)
- Consider using the 'Ptr' version of the 'foo' function here. (V2002)
- Consider using the extended version of the 'foo'function here. (V2001)
- Suspicious subexpression in a sequence of similar comparisons. (V6045)
- The 'Foo' named constant with the value of 0 is used in the bitwise operation. (V616)
- The buffer size is not a multiple of the element size. (V641)
- The constant NN is being utilized. The resulting value could be inaccurate. Consider using the KK constant. (V3026)
- The expression was enclosed by parentheses twice: ((expression)). One pair of parentheses is unnecessary or misprint is present. (V592, V3088)
- The function located to the right of the '|' and '&' operators will be called regardless of the value of the left operand. Consider using '||' and '&&' instead. (V792, V6030)
- The language standard does not define an order in which the 'Foo' functions will be called during evaluation of arguments. (V681)
- The memory allocation and deallocation methods are incompatible. (V611)
- The same sub-expression is present on both sides of the operator. The expression is incorrect or it can be simplified. (V3109)
- The size of the bit mask is less than the size of the first operand. This will cause the loss of the higher bits. (V784)
- The size of the dynamic array can be less than the number of elements in the initializer. (V798)
- The strings were concatenated but are not utilized. Consider inspecting the expression. (V655)
- The value is assigned to the 'x' variable but is not used. (V6021)
- Variables are initialized through the call to the same function. It's probably an error or un-optimized code. (V656, V3086, V6027)
- Character escape is used in multicharacter literal. This causes implementation-defined behavior. (V1039)
- Possible typo in the spelling of a pre-defined macro name. (V1040)
- Class member is initialized with dangling reference. (V1041)
- This file is marked with copyleft license, which requires you to open the derived source code. (V1042)
- A global object variable is declared in the header. Multiple copies of it will be created in all translation units that include this header file. (V1043)
- Loop break conditions do not depend on the number of iterations. (V1044)
- The DllMain function throws an exception. Consider wrapping the throw operator in a try..catch block. (V1045)
- MISRA. Value of the essential character type should be used appropriately in the addition/subtraction operations. (V2543)
- MISRA. The values used in expressions should have appropriate essential types. (V2544)
- MISRA. Conversion between pointers of different object types should not be performed. (V2545)
- MISRA. Expression resulting from the macro expansion should be surrounded by parentheses. (V2546)
- MISRA. The return value of non-void function should be used. (V2547)
- MISRA. The address of an object with local scope should not be passed out of its scope. (V2548)
- MISRA. Pointer to FILE should not be dereferenced. (V2549)
- MISRA. Floating-point values should not be tested for equality or inequality. (V2550)
- MISRA. Variable should be declared in a scope that minimizes its visibility. (V2551)
- MISRA. Expressions with enum underlying type should have values corresponding to the enumerators of the enumeration. (V2552)
- MISRA. Unary minus operator should not be applied to an expression of the unsigned type. (V2553)
- MISRA. Expression containing increment (++) or decrement (--) should not have other side effects. (V2554)
- MISRA. Incorrect shifting expression. (V2555)
- MISRA. Use of a pointer to FILE when the associated stream has already been closed. (V2556)
- MISRA. Operand of sizeof() operator should not have other side effects. (V2557)
- String literal contains potential interpolated expression. (V3138)
- Two or more case-branches perform the same actions. (V3139)
- Property accessors use different backing fields. (V3140)
- Expression under 'throw' is a potential null, which can lead to NullReferenceException. (V3141)
- Unreachable code detected. It is possible that an error is present. (V3142)
- The 'value' parameter is rewritten inside a property setter, and is not used after that. (V3143)
- This file is marked with copyleft license, which requires you to open the derived source code. (V3144)
- Unsafe dereference of a WeakReference target. The object could have been garbage collected before the 'Target' property was accessed. (V3145)
- Two or more case-branches perform the same actions. (V6067)
- Suspicious use of BigDecimal class. (V6068)
- Unsigned right shift assignment of negative 'byte' / 'short' value. (V6069)
- Unsafe synchronization on an object. (V6070)
- This file is marked with copyleft license, which requires you to open the derived source code. (V6071)
- Two similar code fragments were found. Perhaps, this is a typo and 'X' variable should be used instead of 'Y'. (V6072)
